# <center>Car Price Portfolio Project</center>

<div style="text-align: justify"> Cars are one of the transportation that people have. Some people buy cars for daily usage. Some people buy cars because they have good memories from the past. Some even bought a car to complete their collection.
It takes a lot of thought to buy a car, for example, the price. "What's the right price for a car?" To answer this question, I helped by building a machine-learning model that predicts a car's price. </div>

<h2><center>Data Profiling</center></h2>

__What’s data profiling?__
<br>
_Data Profiling is summarizing data using descriptive statistics. Creating data profiling aims to have a strong understanding of data to compile framework analysis and visualize data._

<h3><center>Importing Library</center></h3>

In [437]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

<h3><center>Importing Data</center></h3>

In [438]:
df = pd.read_csv('CarPrice.csv')
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

The first step is initializing the library, importing the dataset into python using **Pandas**, and assigning it as **df**. The data will be saved as **‘CarPrice.csv’**.

<h3><center>Showing The Length of The Data</center></h3>

In [439]:
print("The Length of The Data: ", len(df))

The Length of The Data:  11914


The second step is to **display how much data on the dataset** using **len()**. The result is the size of this data is **11914**.

<h3><center>Showing The Shape of The Data</center></h3>

In [440]:
print("The Shape of The Data: ", df.shape)

The Shape of The Data:  (11914, 16)


The third step is to **display how many rows and columns in the data** use **.shape**. The result is this data has **11914 rows and 16 columns**.

<h3><center>Showing The Information of The Data</center></h3>

In [441]:
print("The Information of The Data: \n", df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

The fourth step is to **get information from the data about what kind of type for each column** using the **.info()**. As you can see, there are three types of data: **object, int64, and float64**.

<h3><center>Showing The Unique Data</center></h3>

In [442]:
print("The The Unique Data:\n", df.nunique())

The The Unique Data:
 Make                   48
Model                 915
Year                   28
Engine Fuel Type       10
Engine HP             356
Engine Cylinders        9
Transmission Type       5
Driven_Wheels           4
Number of Doors         3
Market Category        71
Vehicle Size            3
Vehicle Style          16
highway MPG            59
city mpg               69
Popularity             48
MSRP                 6049
dtype: int64


The fifth step of data profiling is **displaying unique data from each column** using the function **.nunique()**.

<h3><center>Changing The Column's Name</center></h3>

As shown in **Showing The Information of The Data**, the column's name on the dataset is ambiguous. For example, the MSRP column contains the price of a car. Therefore I'm going to change the column's title.

In [443]:
df.rename(columns ={'MSRP': 'Price',
                    'Make': 'Car Brand',
                    'Engine HP': 'Engine Horse Power',
                    'Driven_Wheels':'Driven Mode',
                    'highway MPG':'KML-H',
                    'city mpg':'KML-C'}, inplace=True)

<h2><center>Data Cleansing</center></h2>

After data profiling, I'll do data cleansing. **What's data cleansing?**
<br>
*Data cleansing is identifying incorrect, incomplete, inaccurate, irrelevant, or missing data and then modifying, replacing, or deleting them according to the need.*

<h3><center>Checking The Missing Value in The Car Brand Column</center></h3>

In [444]:
missing_value_Car_Brand = df["Car Brand"].isnull().sum()
if missing_value_Car_Brand > 0:
    print("There are " + str(missing_value_Car_Brand) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The Model Column</center></h3>

In [445]:
missing_value_Model = df["Model"].isnull().sum()
if missing_value_Model > 0:
    print("There are " + str(missing_value_Model) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The Year Column</center></h3>

In [446]:
missing_value_Year = df["Year"].isnull().sum()
if missing_value_Year > 0:
    print("There are " + str(missing_value_Year) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The Engine Fuel Type Column</center></h3>

In [447]:
missing_value_Engine_Fuel_Type = df["Engine Fuel Type"].isnull().sum()
if missing_value_Engine_Fuel_Type > 0:
    print("There are " + str(missing_value_Engine_Fuel_Type) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There are 3 missing values in this column.


In [448]:
print("Here is the data that has a NaN value in the Engine Fuel Type Column:")
df_null_Engine_Fuel_Type = df.loc[df["Engine Fuel Type"].isnull()]
df_null_Engine_Fuel_Type.groupby(["Car Brand", "Model"]).size()

Here is the data that has a NaN value in the Engine Fuel Type Column:


Car Brand  Model 
Suzuki     Verona    3
dtype: int64

The car with the missing value is **Suzuki Verona**. Suzuki Verona uses a **regular unleaded**. So I’m gonna turn this missing value into a ‘regular_unleaded’.

In [449]:
df["Engine Fuel Type"] = np.where((df["Engine Fuel Type"].isnull()) & 
                                  (df["Car Brand"] == "Suzuki") & 
                                  (df['Model'] == 'Verona') & 
                                  (df['Year'] == 2004), 'regular unleaded', df["Engine Fuel Type"])

checking_again_missing_value_Engine_Fuel_Type = df["Engine Fuel Type"].isnull().sum()
if checking_again_missing_value_Engine_Fuel_Type > 0:
    print("There still are " + str(checking_again_missing_value_Engine_Fuel_Type) + " missing values in this column.")
print("After handling the missing values in this column, there is no missing values in this column.")

After handling the missing values in this column, there is no missing values in this column.


<h3><center>Checking The Missing Value in The Engine Horse Power Column</center></h3>

In [450]:
missing_value_Engine_Horse_Power = df["Engine Horse Power"].isnull().sum()
if missing_value_Engine_Horse_Power > 0:
    print("There are " + str(missing_value_Engine_Horse_Power) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There are 69 missing values in this column.


In [451]:
print("Here is the data that has a NaN value in the Engine Horse Power Column:")
df_null_Engine_Horse_Power = df.loc[df["Engine Horse Power"].isnull()]
df_null_Engine_Horse_Power.groupby(["Car Brand", "Model"]).size()

Here is the data that has a NaN value in the Engine Horse Power Column:


Car Brand      Model      
Chevrolet      Impala          6
FIAT           500e            3
Ford           Escape          4
               Focus           3
               Freestar        6
Honda          Fit EV          2
Kia            Soul EV         5
Lincoln        Continental     4
               MKZ             4
Mercedes-Benz  M-Class         1
Mitsubishi     i-MiEV          1
Nissan         Leaf           10
Tesla          Model S        18
Toyota         RAV4 EV         2
dtype: int64

The car with the missing values already mentioned above has a horsepower (HP) as follows:
- Chevrolet Impala's horsepower is 305
- FIAT 500e's horsepower is 111
- Ford Escape's horsepower is 179
- Ford Focus's horsepower is 160
- Ford Freestar's horsepower is 201
- Honda Fit EV's horsepower is 123
- KIA Soul EV's horsepower is 109
- Lincoln Continental's horsepower is 305
- Lincoln MKZ's horsepower is 245
- Mercedes-Benz M-Class's horsepower is 200.
- Mitsubishi i-MiEV's horsepower is 66
- Nissan Leaf's horsepower is 150
- Tesla Model S's horsepower is 1020
- Toyota RAV4 EV's horsepower is 154

In [452]:
df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Chevrolet") & 
                                  (df['Model'] == 'Impala'), 305, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "FIAT") & 
                                  (df['Model'] == '500e'), 111, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Ford") & 
                                  (df['Model'] == 'Escape'), 179, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Ford") & 
                                  (df['Model'] == 'Focus'), 160, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Ford") & 
                                  (df['Model'] == 'Freestar'), 201, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Honda") & 
                                  (df['Model'] == 'Fit EV'), 123, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Kia") & 
                                  (df['Model'] == 'Soul EV'), 109, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Lincoln") & 
                                  (df['Model'] == 'Continental'), 305, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Lincoln") & 
                                  (df['Model'] == 'MKZ'), 245, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Mercedes-Benz") & 
                                  (df['Model'] == 'M-Class'), 200, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Mitsubishi") & 
                                  (df['Model'] == 'i-MiEV'), 66, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Nissan") & 
                                  (df['Model'] == 'Leaf'), 150, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Tesla") & 
                                  (df['Model'] == 'Model S'), 1020, df["Engine Horse Power"])

df["Engine Horse Power"] = np.where((df["Engine Horse Power"].isnull()) & 
                                  (df["Car Brand"] == "Toyota") & 
                                  (df['Model'] == 'RAV4 EV'), 154, df["Engine Horse Power"])


checking_again_missing_value_Engine_Horse_Power = df["Engine Horse Power"].isnull().sum()
if checking_again_missing_value_Engine_Horse_Power > 0:
    print("There still are " + str(checking_again_missing_value_Engine_Horse_Power) + " missing values in this column.")
print("After handling the missing values in this column, there is no missing values in this column.")

After handling the missing values in this column, there is no missing values in this column.


<h3><center>Checking The Missing Value in The Engine Cylinders Column</center></h3>

In [453]:
checking_missing_value_Engine_Cylinders = df["Engine Cylinders"].isnull().sum()
if checking_missing_value_Engine_Cylinders > 0:
    print("There are " + str(checking_missing_value_Engine_Cylinders) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There are 30 missing values in this column.


In [454]:
print("Here is the data that has a NaN value in the Engine Cylinders Column:")
df_null_Engine_Cylinders = df.loc[df["Engine Cylinders"].isnull()]
df_null_Engine_Cylinders.groupby(["Car Brand", "Model"]).size()

Here is the data that has a NaN value in the Engine Cylinders Column:


Car Brand   Model  
Chevrolet   Bolt EV     2
Mazda       RX-7        3
            RX-8       17
Mitsubishi  i-MiEV      3
Toyota      RAV4 EV     1
Volkswagen  e-Golf      4
dtype: int64

From the code above, the cars with missing values are **Chevrolet Bolt EV, Volkswagen E-Golf, Toyota Rav4 EV, Mazda RX-7, and Mazda RX-8**. 
<br>
**Chevrolet Bolt EV, Volkswagen E-Golf, and Toyota Rav4 EV are electric cars**, so it **doesn't have cylinders**. Meanwhile, **Mazda RX-7 and Mazda RX-8 use two cylinders**. 

In [455]:
df["Engine Cylinders"] = np.where((df["Engine Cylinders"].isnull()) & 
                                  (df["Car Brand"] == "Mazda") & 
                                  (df['Model'] == 'RX-7') |
                                  (df['Model'] == 'RX-8'), 2, df["Engine Cylinders"])

df["Engine Cylinders"] = np.where((df["Engine Cylinders"].isnull()), 0, df["Engine Cylinders"])


checking_again_missing_value_Engine_Cylinders = df["Engine Cylinders"].isnull().sum()
if checking_again_missing_value_Engine_Cylinders > 0:
    print("There still are " + str(checking_again_missing_value_Engine_Cylinders) + " missing values in this column.")
print("After handling the missing values in this column, there is no missing values in this column.")

After handling the missing values in this column, there is no missing values in this column.


<h3><center>Checking The Missing Value in The Transmission Type Column</center></h3>

In [456]:
checking_missing_value_Transmission_Type = df["Transmission Type"].isnull().sum()
if checking_missing_value_Transmission_Type > 0:
    print("There are " + str(checking_missing_value_Transmission_Type) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The Driven Mode Column</center></h3>

In [457]:
checking_missing_value_Driven_Mode = df["Driven Mode"].isnull().sum()
if checking_missing_value_Driven_Mode > 0:
    print("There are " + str(checking_missing_value_Driven_Mode) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The Number of Doors Column</center></h3>

In [458]:
checking_missing_value_Number_of_Doors = df["Number of Doors"].isnull().sum()
if checking_missing_value_Number_of_Doors > 0:
    print("There are " + str(checking_missing_value_Number_of_Doors) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There are 6 missing values in this column.


In [459]:
print("Here is the data that has a NaN value in the Number of Doors Column:")
df_null_Engine_Number_of_Doors = df.loc[df["Number of Doors"].isnull()]
df_null_Engine_Number_of_Doors.groupby(["Car Brand", "Model"]).size()

Here is the data that has a NaN value in the Number of Doors Column:


Car Brand  Model  
Ferrari    FF         1
Tesla      Model S    5
dtype: int64

The car with the missing values already mentioned above has a number of doors as follows:
- Ferrari FF's number of doors are 2
- Tesla Model S's number of doors is 4

In [460]:
df["Number of Doors"] = np.where((df["Number of Doors"].isnull()) & 
                                  (df["Car Brand"] == "Tesla") & 
                                  (df['Model'] == 'Model S'), 4, df["Number of Doors"])

df["Number of Doors"] = np.where((df["Number of Doors"].isnull()) & 
                                  (df["Car Brand"] == "Ferrari") & 
                                  (df['Model'] == 'FF'), 2, df["Number of Doors"])


checking_again_missing_value_Number_of_Doors = df["Number of Doors"].isnull().sum()
if checking_again_missing_value_Number_of_Doors > 0:
    print("There still are " + str(checking_again_missing_value_Number_of_Doors) + " missing values in this column.")
print("After handling the missing values in this column, there is no missing values in this column.")

After handling the missing values in this column, there is no missing values in this column.


<h3><center>Checking The Missing Value in The Market Category Column</center></h3>

In [461]:
checking_missing_value_Market_Category = df["Market Category"].isnull().sum()
if checking_missing_value_Market_Category > 0:
    print("There are " + str(checking_missing_value_Market_Category) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There are 3742 missing values in this column.


In [462]:
df = df.drop(['Market Category'], axis=1)

In [463]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Car Brand           11914 non-null  object 
 1   Model               11914 non-null  object 
 2   Year                11914 non-null  int64  
 3   Engine Fuel Type    11914 non-null  object 
 4   Engine Horse Power  11914 non-null  float64
 5   Engine Cylinders    11914 non-null  float64
 6   Transmission Type   11914 non-null  object 
 7   Driven Mode         11914 non-null  object 
 8   Number of Doors     11914 non-null  float64
 9   Vehicle Size        11914 non-null  object 
 10  Vehicle Style       11914 non-null  object 
 11  KML-H               11914 non-null  int64  
 12  KML-C               11914 non-null  int64  
 13  Popularity          11914 non-null  int64  
 14  Price               11914 non-null  int64  
dtypes: float64(3), int64(5), object(7)
memory usage: 1.4+

<h3><center>Checking The Missing Value in The Vehicle Size Column</center></h3>

In [464]:
checking_missing_value_Vehicle_Size = df["Vehicle Size"].isnull().sum()
if checking_missing_value_Vehicle_Size > 0:
    print("There are " + str(checking_missing_value_Market_Category) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The Vehicle Style Column</center></h3>

In [465]:
checking_missing_value_Vehicle_Style = df["Vehicle Style"].isnull().sum()
if checking_missing_value_Vehicle_Style > 0:
    print("There are " + str(checking_missing_value_Vehicle_Style) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The KML-H Column</center></h3>

In [466]:
checking_missing_value_KML_H = df["KML-H"].isnull().sum()
if checking_missing_value_KML_H > 0:
    print("There are " + str(checking_missing_value_KML_H) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


Changing the KML-H values, from mile per gallon to km per liter.

In [467]:
df['KML-H'] = df['KML-H'].apply(lambda x: round(x*0.425144))

<h3><center>Checking The Missing Value in The KML-C Column</center></h3>

In [468]:
checking_missing_value_KML_C = df["KML-C"].isnull().sum()
if checking_missing_value_KML_C > 0:
    print("There are " + str(checking_missing_value_KML_C) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


Changing the KML-C values, from mile per gallon to km per liter.

In [469]:
df['KML-C'] = df['KML-C'].apply(lambda x: round(x*0.425144))

<h3><center>Checking The Missing Value in The Popularity Column</center></h3>

In [470]:
checking_missing_value_Popularity = df["Popularity"].isnull().sum()
if checking_missing_value_Popularity > 0:
    print("There are " + str(checking_missing_value_Popularity) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h3><center>Checking The Missing Value in The Price Column</center></h3>

In [471]:
checking_missing_value_Price = df["Price"].isnull().sum()
if checking_missing_value_Price > 0:
    print("There are " + str(checking_missing_value_Price) + " missing values in this column.")
else:
    print("There is no missing values in this column.")

There is no missing values in this column.


<h2><center>Data Exploration</center></h2>

In [472]:
df.head()

Car Brand       Model  Year             Engine Fuel Type  \
0       BMW  1 Series M  2011  premium unleaded (required)   
1       BMW    1 Series  2011  premium unleaded (required)   
2       BMW    1 Series  2011  premium unleaded (required)   
3       BMW    1 Series  2011  premium unleaded (required)   
4       BMW    1 Series  2011  premium unleaded (required)   

   Engine Horse Power  Engine Cylinders Transmission Type       Driven Mode  \
0               335.0               6.0            MANUAL  rear wheel drive   
1               300.0               6.0            MANUAL  rear wheel drive   
2               300.0               6.0            MANUAL  rear wheel drive   
3               230.0               6.0            MANUAL  rear wheel drive   
4               230.0               6.0            MANUAL  rear wheel drive   

   Number of Doors Vehicle Size Vehicle Style  KML-H  KML-C  Popularity  Price  
0              2.0      Compact         Coupe     11      8        3916  46135  
1              2.0      Compact   Convertible     12      8        3916  40650  
2              2.0      Compact         Coupe     12      9        3916  36350  
3              2.0      Compact         Coupe     12      8        3916  29450  
4              2.0      Compact   Convertible     12      8        3916  34500

<h3><center>Checking How Many Car Brand in The Dataset</center></h3>

In [473]:
print("Here are the TOP TEN of Car Brands that exist in this dataset:")
df_Car_Brand = df.groupby(["Car Brand"]).size().sort_values(ascending=False)
df_Car_Brand.head(10)

Here are the TOP TEN of Car Brands that exist in this dataset:


Car Brand
Chevrolet     1123
Ford           881
Volkswagen     809
Toyota         746
Dodge          626
Nissan         558
GMC            515
Honda          449
Mazda          423
Cadillac       397
dtype: int64

<h3><center>Checking How Many Car Brand With The Model in The Dataset</center></h3>

In [474]:
print("Here are the TOP TEN of Car Brands with the Model that exist in this dataset:")
df_Car_Brand_Model = df.groupby(["Car Brand", "Model"]).size().sort_values(ascending=False)
df_Car_Brand_Model.head(10)

Here are the TOP TEN of Car Brands with the Model that exist in this dataset:


Car Brand   Model             
Chevrolet   Silverado 1500        156
Toyota      Tundra                140
Ford        F-150                 126
GMC         Sierra 1500            90
Volkswagen  Beetle Convertible     89
Toyota      Tacoma                 80
Nissan      Frontier               76
Volkswagen  GTI                    76
Honda       Accord                 75
Volkswagen  Beetle                 75
dtype: int64